File: aoc2021/Day15.py
import heapq
from collections import defaultdict
from math import inf as INFINITY

def dijkstra(grid):
    height, width = len(grid), len(grid[0])
    start = (0,0)
    end = (height - 1, width - 1)

    queue = [(0, start)]

    minDist = defaultdict(lambda:INFINITY, {start: 0})
    visited = set()

    while queue:
        distance, node = heapq.heappop(queue)

        if node == end:
            return distance

        if node in visited:
            continue

        visited.add(node)
        row, col = node

        for neighbor in neighbors4(row, col, height, width):
            if neighbor in visited:
                continue

            nextRow, nextCol = neighbor
            newDistance = distance + grid[nextRow][nextCol]
            
            if newDistance < minDist[neighbor]:
                minDist[neighbor] = newDistance
                heapq.heappush(queue, (newDistance, neighbor))

    return INFINITY
    

def neighbors4(row, col, height, width):
    directions = ((1,0),(-1,0),(0,1),(0,-1))

    for dirRow, dirCol in directions:
        rowIterator, colIterator = (row + dirRow, col + dirCol)
        if 0 <= rowIterator < width and 0 <= colIterator < height:
            yield rowIterator, colIterator
        

def main():
    with open('C:\dev\Advent of Code\Advent of Code 2021\Day 15\inputDay15.txt') as file:
        input = map(str.rstrip, file) 
        riskMap = list(list(map(int, row)) for row in input)

    print(dijkstra(riskMap))

if __name__ == '__main__':
    main()

File: aoc2021/Day 15.cpp
#include <iostream>
#include <math.h>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>
#include <bits/stdc++.h>

class Part1 {
public:
    void dijkstra(std::vector<std::vector<int>> riskMap, std::vector<std::vector<bool>>& isVisited, std::vector<std::vector<int>>& minDist);
};

int main() {
    std::ifstream file("inputDay15.txt");

    std::string line;
    std::string num;
    std::vector<std::vector<int>> riskMap;
    std::vector<std::vector<bool>> isVisited;
    std::vector<std::vector<int>> minDist;

    while (getline(file, line)) {
        std::stringstream ss(line);
        std::vector<int> riskLine;
        std::vector<int> costLine;
        std::vector<bool> isVisitedLine;
        getline(ss, num);
        for (int i = 0; i < num.size(); i++) {
            riskLine.push_back(num[i] - '0');
            isVisitedLine.push_back(false);
            costLine.push_back(1000000);
        }

        riskMap.push_back(riskLine);
        isVisited.push_back(isVisitedLine);
        minDist.push_back(costLine);
    }
    file.close();
    minDist[0][0] = 0;

    // for (auto&& line : riskMap) {
    //     for (auto&& i : line) {
    //         std::cout << i;
    //     }
    // }

    Part1 solutiuon;

    solutiuon.dijkstra(riskMap, isVisited, minDist);

    return 0;
}

void Part1::dijkstra(std::vector<std::vector<int>> riskMap, std::vector<std::vector<bool>>& isVisited, std::vector<std::vector<int>>& minDist) {
    int h = riskMap.size();
    int w = riskMap[0].size();

    std::pair<int, int> start(0, 0);
    // std::vector<std::pair<int, std::pair<int, int>>> queue;
    // std::priority_queue<int, std::pair<int, int>> queue;
    // queue.push(0, (start));
    std::priority_queue<int, int> queue;
    queue.push(0, 0);


    // while (queue) {
    //     distance = queue.pop
    // }




}
